### PR TITLE
Source postgres: Add SSL certificates to source postgres

### LINKED ISSUE #4034: Support Postgres SSL connection
Currently the Postgres connector has an SSL toggle, however there is no where to specify the appropriate information to establish the SSL connection:
- CA certificate.
- Client certificate.
- Client key.
- Client key password.
- SSL Mode.

This is a must for us a all the connections to our Postgres databases require SSL.

### LINKED ISSUE #12766: Update acceptance tests for Postgres source/destination with SSL certificates
At this time, we enable SSL connections using a boolean switch and we do not need certificates to configure the connection.

Once the changes from Task - https://github.com/airbytehq/airbyte/issues/4034 are implemented, we will need both client and server certificates to test the SSL connection. This requires creating such certificates in the test container, disabling the SSL connection and updating existing ones, and adding new Acceptance Tests for postgres source and destgnestion.

### PR DESCRIPTION
**Pull request by andriikorotkov**
## What
Add SSL certificates to source and destination postgres

## Recommended reading order
1. `PostgresSource.java`
2. `PostgresDestination.java`

## 🚨 User Impact 🚨
No impact

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/destination-postgres/src/main/java/io/airbyte/integrations/destination/postgres/PostgresDestination.java
Comment (alexandr-shegeda) [inline airbyte-integrations/connectors/destination-postgres/src/main/java/io/airbyte/integrations/destination/postgres/PostgresDestination.java:108]: does it make sense to move logic around the certificates into a separate utility class?

Comment (andriikorotkov) [reply airbyte-integrations/connectors/destination-postgres/src/main/java/io/airbyte/integrations/destination/postgres/PostgresDestination.java]: Yes of course. I transferred the general logic to a separate class.

File: airbyte-integrations/connectors/destination-postgres/src/main/java/io/airbyte/integrations/destination/postgres/PostgresDestination.java
Comment (alexandr-shegeda) [inline airbyte-integrations/connectors/destination-postgres/src/main/java/io/airbyte/integrations/destination/postgres/PostgresDestination.java:52]: just want to confirm that my understanding is correct, in case SSL_KEY is not provided we are going to enable SSL by default, is it expected behavior?

Comment (andriikorotkov) [reply airbyte-integrations/connectors/destination-postgres/src/main/java/io/airbyte/integrations/destination/postgres/PostgresDestination.java:52]: yes, this is normal behavior, which I modified for different types of SSL mode. I will add tests that test this.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/util/PostgresSslConnectionUtils.java
Comment (mkhokh-33) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/util/PostgresSslConnectionUtils.java:142]: i guess in other connector we have 30 seconds for timeout, could you pls check ? :)

**Review by mkhokh-33 — CHANGES_REQUESTED** [review]

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/util/PostgresSslConnectionUtils.java
Comment (andriikorotkov) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/util/PostgresSslConnectionUtils.java:142]: updated it

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/util/PostgresSslConnectionUtils.java
Comment (andriikorotkov) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/util/PostgresSslConnectionUtils.java:30]: updated it

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/util/PostgresSslConnectionUtils.java
Comment (andriikorotkov) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/util/PostgresSslConnectionUtils.java:46]: updated it

**Review by mkhokh-33 — APPROVED** [review]

**Review by mkhokh-33 — APPROVED** [review]

**Comment by grishick** [comment]
@andriikorotkov have you tested these changes with "strict-encrypt" connectors on cloud?

File: airbyte-integrations/connectors/destination-postgres/src/main/resources/spec.json
Comment (grishick) [inline airbyte-integrations/connectors/destination-postgres/src/main/resources/spec.json:71]: nit: make the titles lowercase, so that they match the `sslmode` values and the description, which mentions them in lower case.

File: airbyte-integrations/connectors/destination-postgres-strict-encrypt/src/test/resources/expected_spec.json
Comment (grishick) [inline airbyte-integrations/connectors/destination-postgres-strict-encrypt/src/test/resources/expected_spec.json:59]: Looks like these descriptions are copied from Postgres docs, which is OK, but I prefer to not copy/paste from external sources and I'd like us to more user friendly descriptions:
- change "SSL connection configuration modes" to "SSL connection modes"
- change description of "disable" to "Chose this mode to disable encryption of communication between Airbyte and source database"
- change description of "allow" to "Chose this mode to enable encryption *only* when required by the source database"
- change description of "prefer" to "Chose this mode to allow unencrypted connection *only* if the source database does not support encryption"
- change description of "require" to "Chose this mode to always require encryption. If the source database server does not support encryption, connection will fail"
- change description of "verify-ca" to "Chose this mode to always require encryption and to verify that the source database server has a valid SSL certificate"
- change description of "verify-full" to "This is the most secure mode. Chose this mode to always require encryption and to verify the identity of the source database server"

**Review by grishick — COMMENTED** [review]
I left a couple of minor comments about the spec. In addition to integration testing, I need you to perform the following manual test:

1. Create a connection with the previous version of source-postgres-strict-encrypt connector and run a full refresh (destination does not matter as long as the connection is configured for 'incremental dedupe' mode)
2. Add some data to the source DB and run an incremental sync
3. Upgrade source-postgres-strict-encrypt to a version with these changes. Don't change any configuration.
4. Run an incremental sync
5. Add some data to the source DB 
6. Downgrade the connector to the previous version
7. Run an incremental sync
8. Upgrade source-postgres-strict-encrypt to a version with these changes.
9. Change SSL mode to require
10. Add some data to the source DB 
11. Run an incremental sync
12. Downgrade the connector to the previous version
13. Add some data to the source DB 
14. Run an incremental sync

The expectation for the test is that all syncs succeed

File: airbyte-integrations/connectors/destination-postgres/src/main/resources/spec.json
Comment (andriikorotkov) [reply airbyte-integrations/connectors/destination-postgres/src/main/resources/spec.json:71]: thanks, I changed the titles of all modes in the specifications of the postgres connectors

**Review by edgao — COMMENTED** [review]
We also need to update normalization to handle these configs, right?

**Comment by edgao** [comment]
> @grishick, I see this problem with almost all the connectors we have. I created a new issue so that this could be fixed in the complex. Here is the link to the issue - #14345.

@andriikorotkov is that the same problem as https://github.com/airbytehq/airbyte/issues/14196 ?

**Comment by grishick** [comment]
> I can build an intermediate version of postgres controllers with this changed setting (just for this) and then add my changes with SSL modes. Otherwise, we need to ignore it and users who have downgraded the connector version will have to save the specification again (I can add a comment to the documentation about this).

This seems like a good idea until we get rid of `additionalProperties` altogether.

**Comment by andriikorotkov** [comment]
> 

@edgao, Yes you're right. This is one and the same issue. Should I close my issue? Or is it better to add a link to my task in your issue?

**Comment by andriikorotkov** [comment]
> > @grishick, I see this problem with almost all the connectors we have. I created a new issue so that this could be fixed in the complex. Here is the link to the issue - #14345.
> 
> @andriikorotkov is that the same problem as #14196 ?

@edgao, Yes you're right. This is one and the same issue. Should I close my issue? Or is it better to add a link to my task in your issue?

**Comment by andriikorotkov** [comment]
> We also need to update normalization to handle these configs, right?

I think we can leave normalization without updates. It works fine with the tests I added, and I also tested it locally when I manually tested these changes.

File: airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java
Comment (andriikorotkov) [reply airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java:119]: It is not quite so. if the user has not previously used an SSL connection (the SSL toggle was disabled), then SSL will be disabled with the updated version. But, if the user used an SSL connection (the SSL toggle was turned on) - then when changing the version, the SSL connection will be turned on, and will work in the mode that was in the old version (in the old versions, when turning on the SSL connection, we added SSL mode = require on the java part)

File: airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSourceStrictEncrypt.java
Comment (andriikorotkov) [reply airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSourceStrictEncrypt.java:38]: Thank you, this is really much better than my loop addition. Changed it.

File: airbyte-integrations/connectors/source-postgres/src/main/resources/spec.json
Comment (andriikorotkov) [reply airbyte-integrations/connectors/source-postgres/src/main/resources/spec.json:72]: This text is for the user to have a description of SSL modes and to be able to link quickly to the documentation. I am attaching a screenshot of how it looks to the user.

File: airbyte-db/db-lib/src/main/java/io/airbyte/db/PostgresUtils.java
Comment (andriikorotkov) [reply airbyte-db/db-lib/src/main/java/io/airbyte/db/PostgresUtils.java:31]: well, I'm fine with `Triple(...)` =), but it's a good idea to make an inner class so developers reading this code can better understand what's what. Changed it.

File: airbyte-integrations/connectors/source-postgres/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/PostgresSourceSSLCaCertificateAcceptanceTest.java
Comment (andriikorotkov) [reply airbyte-integrations/connectors/source-postgres/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/PostgresSourceSSLCaCertificateAcceptanceTest.java:65]: Yes, of course it works correctly. In order for `SSL mode = verify-ca` to work well, you need to add a root certificate to the server and restart the server. I did it in the settings so it should work fine.

File: airbyte-integrations/connectors/source-postgres/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/PostgresSourceStrictEncryptAcceptanceTest.java
Comment (andriikorotkov) [reply airbyte-integrations/connectors/source-postgres/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/PostgresSourceStrictEncryptAcceptanceTest.java:50]: If you mean everything marked in the `withCommand` parameter, then yes, this functionality has moved to the `getCertificate()` method. Also, I'm using the official postgres docker container, instead of the custom one that marcosmarxm created.

File: airbyte-integrations/connectors/source-postgres/src/test/java/io/airbyte/integrations/source/postgres/AbstractPostgresSourceSSLCertificatesTest.java
Comment (andriikorotkov) [reply airbyte-integrations/connectors/source-postgres/src/test/java/io/airbyte/integrations/source/postgres/AbstractPostgresSourceSSLCertificatesTest.java:175]: You noted it well. Thanks, I removed this method and added the rows for deleting keys to the @AfterAll method.

File: airbyte-db/db-lib/src/main/java/io/airbyte/db/PostgresUtils.java
Comment (andriikorotkov) [reply airbyte-db/db-lib/src/main/java/io/airbyte/db/PostgresUtils.java:32]: Sorry, but for some reason my IDEA can't accept this as text. Instead, I get the following:


As for combining everything into one command, it won't work, because each command must be executed with SUDO rights, which I do at the beginning of each command. Also, if I do that, it will get mixed up and look unreadable. So it's best to leave it split into separate commands.

**Comment by andriikorotkov** [comment]
@grishick @edgao, please, if everything is good in this pull request and you like everything, please give your approval. So that we do not postpone these changes for a long time.

File: airbyte-integrations/connectors/source-postgres/src/test/java/io/airbyte/integrations/source/postgres/AbstractPostgresSourceSSLCertificatesTest.java
Comment (andriikorotkov) [reply airbyte-integrations/connectors/source-postgres/src/test/java/io/airbyte/integrations/source/postgres/AbstractPostgresSourceSSLCertificatesTest.java]: @edgao, I'm sorry, but I need to delete certificates after each unit test. For acceptance tests - I removed it, but for unit tests - I put this method back.

**Comment by edgao** [comment]
> > > @grishick, I see this problem with almost all the connectors we have. I created a new issue so that this could be fixed in the complex. Here is the link to the issue - #14345.
> > 
> > 
> > @andriikorotkov is that the same problem as #14196 ?
> 
> @edgao, Yes you're right. This is one and the same issue. Should I close my issue? Or is it better to add a link to my task in your issue?

let's just link them

---

> > We also need to update normalization to handle these configs, right?
> 
> I think we can leave normalization without updates. It works fine with the tests I added, and I also tested it locally when I manually tested these changes.

from https://github.com/airbytehq/airbyte/issues/4034#issuecomment-1054491496

> our use case is connecting to a Heroku private space DB, which allows for external connections using certificates.

do we have the ability to test this? It sounds like currently their infra would completely reject our connection. I would expect normalization to need the same SSL behavior as the destination.

I think (?) dbt implemented this feature last December (https://github.com/dbt-labs/dbt-core/issues/3472)

File: airbyte-db/db-lib/src/main/java/io/airbyte/db/PostgresUtils.java
Comment (edgao) [reply airbyte-db/db-lib/src/main/java/io/airbyte/db/PostgresUtils.java:31]: If you open the project structure settings (`cmd`-`;` on macos, not sure on windows) you can set the language level to 17:

Screen Shot 2022-07-05 at 3 36 37 PM

which should enable multiline strings and record classes.

File: airbyte-integrations/connectors/source-postgres/src/main/resources/spec.json
Comment (edgao) [reply airbyte-integrations/connectors/source-postgres/src/main/resources/spec.json:72]: neat! thanks for confirming :+1:

**Review by edgao — COMMENTED** [review]
code looks good, just one small fyi about java language level.

approval pending discussion around normalization - I'm definitely not an expert on ssh, so feel free to tell me I have no idea what I'm talking about :P

**Comment by grishick** [comment]
@andriikorotkov please break this PR down into two: one for destination and another one for source.

**Comment by andriikorotkov** [comment]
> @andriikorotkov please break this PR down into two: one for destination and another one for source.

@grishick no problem, but I have to do these pull requests one by one, because the tests will use the shared functionality. So as not to clean up conflicts once again. I'm leaving the postgres source in this pull request, and after it's published and merged, I'll add a pull request with the destination postgres.

**Comment by grishick** [comment]
> > @andriikorotkov please break this PR down into two: one for destination and another one for source.
> 
> @grishick no problem, but I have to do these pull requests one by one, because the tests will use the shared functionality. So as not to clean up conflicts once again. I'm leaving the postgres source in this pull request, and after it's published and merged, I'll add a pull request with the destination postgres.

Sounds good. You can rename this PR and leave on source-related changes here and once this is merged you can create a PR for the destination.

**Comment by grishick** [comment]
One more thing, since Postgres Source is now in beta, please do not publish a new version until the PR for Postgres Source is approved.

**Comment by andriikorotkov** [comment]
> One more thing, since Postgres Source is now in beta, please do not publish a new version until the PR for Postgres Source is approved.

Yes of course. Without approval from you or someone from the Airbyte team, I will not publish a new version of either source or destination.

**Comment by andriikorotkov** [comment]
@edgao As Greg and I discussed above, I split this pull request into two. In this pull request, I have left the changes for the postgres source, and after those changes are published and merged, I will create a new pull request with the changes for the postgres destination. Therefore, changes to normalization will be relevant in the next pull request. I will definitely add you to it as a reviewer.

**Comment by grishick** [comment]
Changes look good. However, to be safe, we should close [this issue](https://github.com/airbytehq/airbyte/issues/14196) before publishing and merging this.

**Review by grishick — APPROVED** [review]
please rebase before publishing and merging

**Comment by andriikorotkov** [comment]
/publish connector=connectors/source-postgres-strict-encrypt auto-bump-version=false

**Comment by VitaliiMaltsev** [comment]
https://github.com/airbytehq/airbyte/pull/14574 was merged. Please proceed

**Pull request merged**
